spec/compiler/interpreter/multidispatch_spec.cr

```diff
@@ -393,5 +393,34 @@ describe Crystal::Repl::Interpreter do
         ("" || nil).foo 1, file: ""
       CODE
     end
+
+    it "does multidispatch with captured block (#12217)" do
+      interpret(<<-CODE).should eq(42)
+        class A
+          def then(&callback : Int32 -> Int32)
+            callback.call(10)
+          end
+        end
+
+        class B < A
+          def then(&callback : Int32 -> Int32)
+            callback.call(30)
+          end
+        end
+
+        a = A.new
+        b = B.new
+
+        a_value = (a || b).then do |x|
+          x + 1
+        end
+
+        b_value = (b || a).then do |x|
+          x + 1
+        end
+
+        a_value + b_value
```

Comment:
Due to the laws of commutativity, this spec would probably pass even if multi-dispatch selected exactly the opposite target def (`b` for `a || b` and `a` for `b || a`).
So I'd suggest to make the results clearly identifyable, for example with this (and expected value `1131`).
```suggestion
        a_value * 100 + b_value
```

Reply:
Great point! I used subtraction instead.